examples/dynamic-bridge-app/linux/include/DataModel.h

```diff
@@ -0,0 +1,218 @@
+        uint16_t index = aPath.mListIndex.Value();
+        if (index >= x.size())
+            return CHIP_ERROR_INVALID_ARGUMENT;
+
+        return aEncoder.Encode(x.data()[index]);
+    }
+    return aEncoder.EncodeList([x](const auto & encoder) {
+        CHIP_ERROR err = CHIP_NO_ERROR;
+        for(auto& v : x) {
+            err = encoder.Encode(v);
+            if (err != CHIP_NO_ERROR)
+                break;
+        }
+        return err;
+    });
+}
+
+/*
+ * @brief
+ * Set of overloaded encode methods that can be called from AttributeAccessInterface::Read
+ */
+template <typename X>
+CHIP_ERROR Encode(const ConcreteReadAttributePath& aPath, AttributeValueEncoder & aEncoder, const Optional<X> & x)
+{
+    if (x.HasValue())
+    {
+        return Encode(aPath, aEncoder, x.Value());
+    }
+    // If no value, just do nothing.
+    return CHIP_NO_ERROR;
+}
+
+/*
+ * @brief
+ *
+ * Encodes a nullable value.
+ */
+template <typename X>
+CHIP_ERROR Encode(const ConcreteReadAttributePath& aPath, AttributeValueEncoder & aEncoder, const Nullable<X> & x)
+{
+    if (x.IsNull())
+    {
+        return aEncoder.EncodeNull();
+    }
+
+    // Allow sending invalid values for nullables when
+    // CONFIG_BUILD_FOR_HOST_UNIT_TEST is true, so we can test how the other side
+    // responds.
+#if !CONFIG_BUILD_FOR_HOST_UNIT_TEST
+    if (!x.HasValidValue())
+    {
+        return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+    }
+#endif // !CONFIG_BUILD_FOR_HOST_UNIT_TEST
+
+    // The -Wmaybe-uninitialized warning gets confused about the fact
+    // that x.mValue is always initialized if x.IsNull() is not
+    // true, so suppress it for our access to x.Value().
+#pragma GCC diagnostic push
+#if !defined(__clang__)
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif // !defined(__clang__)
+    return Encode(aPath, aEncoder, x.Value());
+#pragma GCC diagnostic pop
+}
+
+
+template<typename X, std::enable_if_t<!IsOptionalOrNullable<std::decay_t<X>>::value && !IsList<std::decay_t<X>>::value, bool> = true>
+CHIP_ERROR Decode(const ConcreteDataAttributePath&, AttributeValueDecoder & aDecoder, X& x)
+{
+    return aDecoder.Decode(x);
+}
+
+/*
+ * @brief
+ * Lists that are string-like should be decoded as char/byte spans.
+ */
+template<typename X, std::enable_if_t<IsList<std::decay_t<X>>::value && sizeof(std::decay_t<typename X::pointer>) == sizeof(char), bool> = true>
+CHIP_ERROR Decode(const ConcreteDataAttributePath& aPath, AttributeValueDecoder & aDecoder, X& x)
+{
+    Span<std::decay_t<typename X::pointer>> span;
+    CHIP_ERROR err = aDecoder.Decode(span);
+    if (err = CHIP_NO_ERROR) {
+        x = X(span.data(), span.size());
+    }
+    return err;
+}
+
+
+template<typename X, std::enable_if_t<IsList<std::decay_t<X>>::value && (sizeof(std::decay_t<typename X::pointer>) > sizeof(char)), bool> = true>
+CHIP_ERROR Decode(const ConcreteDataAttributePath& aPath, AttributeValueDecoder & aDecoder, X& x)
+{
+    switch(aPath.mListOp) {
```

Comment:
What about the list ReplaceAll operation?

Reply:
Not sure it's possible to do here without extra state since we would just get decode, decode, decode, decode and none of them would know here if it's the first one. Unless that's set for only one?

Reply:
I'm talking about dealing with `ConcreteDataAttributePath::ListOperation::ReplaceAll`.

Reply:
It looks like it always replaces with a single element? That's doable